### ISSUE TITLE
New rule: UnnecesaryOptInAnnotation

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#5889]
[Kotlin has an `Opt-in` mechanism](https://kotlinlang.org/docs/opt-in-requirements.html) to make sure that the users know that they are using a not stable API. They used on their Standard library and it's used by third parties too (even detekt uses it):

https://github.com/detekt/detekt/blob/4a89a25c82d35874571270fa1ff32751f2f268a1/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/UnstableApi.kt#L3-L7

This is a great mechanism but when an API go from unstable to stable a lot of `@OptIn` annotations are left back on the code. Also this happen when later you refactor the code and remove the usage of that unstable api.

My proposal is to implement a rule to check for this cases so we don't have any unused `@OptIn` annotation in the code.

I see warnings on IntelliJ so maybe we could port it.

It would be great to also raise usages of propagating opt-in but I'm not sure if we could find those cases.

**Comment by mustafaozhan** [comment]
Hello,

This Rule will be really nice to have, I would love to see it.

Meanwhile, I want to share my experience with `@OptIn`. While ago, I add this block to my codebase:
```kotlin
    tasks.withType<KotlinCompile> {
        kotlinOptions {
            allWarningsAsErrors = true
        }
    }
```
In order to fail the CI when there is a warning. Then I realized that `@OptIn` annotations are acted as warning too. After a bit of searching, I replaced my all `@OptIn` with `@Suppress("OPT_IN_USAGE")`. I think when we implement this rule, we should also consider the same rule for  `@Suppress("OPT_IN_USAGE")`.

**Comment by BraisGabin** [comment]
> I believe the Kotlin compiler is also raising a warning in this case right?

As far as I know the compiler doesn't raise any warning. Intellij marks it as unused, but the compiler doesn't complain.
